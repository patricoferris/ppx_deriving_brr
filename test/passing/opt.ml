module Shape : sig
  type t
  type opt

  val opt :
    ?width:int -> ?height:int -> ?two_d:bool -> urls:string array -> unit -> opt

  val create : ?opt:opt -> string -> t
end = struct
  type t = Jv.t

  type opt_internal = {
    width : int option;
    height : int option;
    two_d : bool option;
    urls : string array;
  }
  [@@deriving brr_opt]

  type opt = Jv.t

  let opt = opt_internal_to_jv

  let create ?(opt = Jv.undefined) s =
    Jv.new' (Jv.obj [||]) [| Jv.of_string s; opt |]
end
