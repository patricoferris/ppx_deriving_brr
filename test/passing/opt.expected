module Shape :
  sig
    type t
    type opt
    val opt : ?width:int -> ?height:int -> ?two_d:bool -> unit -> opt
    val create : ?opt:opt -> string -> t
  end =
  struct
    type t = Jv.t
    type opt_internal =
      {
      width: int option ;
      height: int option ;
      two_d: bool option }[@@deriving brr_opt]
    include
      struct
        let opt_internal_to_jv ?width  ?height  ?two_d  () =
          let jv = Jv.obj [||] in
          ((function | None -> () | Some two_d -> Jv.Bool.set jv "twoD" two_d))
            two_d;
          ((function
            | None -> ()
            | Some height -> Jv.Int.set jv "height" height)) height;
          ((function | None -> () | Some width -> Jv.Int.set jv "width" width))
            width;
          jv
      end[@@ocaml.doc "@inline"][@@merlin.hide ]
    type opt = Jv.t
    let opt = opt_internal_to_jv
    let create ?(opt= Jv.undefined)  s =
      Jv.new' (Jv.obj [||]) [|(Jv.of_string s);opt|]
  end 
